cpp/src/arrow/flight/sql/odbc/flight_sql/utils.cc

```diff
@@ -0,0 +1,1124 @@
+      return arrow::Type::INT8;
+    case odbcabstraction::CDataType_UTINYINT:
+      return arrow::Type::UINT8;
+    case odbcabstraction::CDataType_SBIGINT:
+      return arrow::Type::INT64;
+    case odbcabstraction::CDataType_UBIGINT:
+      return arrow::Type::UINT64;
+    case odbcabstraction::CDataType_BINARY:
+      return arrow::Type::BINARY;
+    case odbcabstraction::CDataType_NUMERIC:
+      return arrow::Type::DECIMAL128;
+    case odbcabstraction::CDataType_TIMESTAMP:
+      return arrow::Type::TIMESTAMP;
+    case odbcabstraction::CDataType_TIME:
+      return arrow::Type::TIME64;
+    case odbcabstraction::CDataType_DATE:
+      return arrow::Type::DATE64;
+    default:
+      throw odbcabstraction::DriverException(std::string("Invalid target type: ") +
+                                             std::to_string(data_type));
+  }
+}
+
+odbcabstraction::CDataType ConvertArrowTypeToC(arrow::Type::type type_id,
+                                               bool useWideChar) {
+  switch (type_id) {
+    case arrow::Type::STRING:
+      return GetDefaultCCharType(useWideChar);
+    case arrow::Type::INT16:
+      return odbcabstraction::CDataType_SSHORT;
+    case arrow::Type::UINT16:
+      return odbcabstraction::CDataType_USHORT;
+    case arrow::Type::INT32:
+      return odbcabstraction::CDataType_SLONG;
+    case arrow::Type::UINT32:
+      return odbcabstraction::CDataType_ULONG;
+    case arrow::Type::FLOAT:
+      return odbcabstraction::CDataType_FLOAT;
+    case arrow::Type::DOUBLE:
+      return odbcabstraction::CDataType_DOUBLE;
+    case arrow::Type::BOOL:
+      return odbcabstraction::CDataType_BIT;
+    case arrow::Type::INT8:
+      return odbcabstraction::CDataType_STINYINT;
+    case arrow::Type::UINT8:
+      return odbcabstraction::CDataType_UTINYINT;
+    case arrow::Type::INT64:
+      return odbcabstraction::CDataType_SBIGINT;
+    case arrow::Type::UINT64:
+      return odbcabstraction::CDataType_UBIGINT;
+    case arrow::Type::BINARY:
+      return odbcabstraction::CDataType_BINARY;
+    case arrow::Type::DECIMAL128:
+      return odbcabstraction::CDataType_NUMERIC;
+    case arrow::Type::DATE64:
+    case arrow::Type::DATE32:
+      return odbcabstraction::CDataType_DATE;
+    case arrow::Type::TIME64:
+    case arrow::Type::TIME32:
+      return odbcabstraction::CDataType_TIME;
+    case arrow::Type::TIMESTAMP:
+      return odbcabstraction::CDataType_TIMESTAMP;
+    default:
+      throw odbcabstraction::DriverException(std::string("Invalid type id: ") +
+                                             std::to_string(type_id));
+  }
+}
+
+std::shared_ptr<arrow::Array> CheckConversion(const arrow::Result<arrow::Datum>& result) {
+  if (result.ok()) {
+    const arrow::Datum& datum = result.ValueOrDie();
+    return datum.make_array();
+  } else {
+    throw odbcabstraction::DriverException(result.status().message());
+  }
+}
+
+ArrayConvertTask GetConverter(arrow::Type::type original_type_id,
```

Comment:
It seems this should take a full DataType and not just the type ID? In the source it is called in a context where it has the full type. That way you don't need GetDefaultDataTypeForTypeId (Also, maybe it should be moved to flight_sql_result_set_column.cc since it is only used there?)